lua/sim/MatchState.lua

```diff
@@ -0,0 +1,209 @@
+
+local Conditions = {
+    demoralization = categories.COMMAND,
+    domination = categories.STRUCTURE + categories.ENGINEER - categories.WALL,
+    eradication = categories.ALLUNITS - categories.WALL,
+}
+
+--- Ends the game
+function CallEndGame()
+    ForkThread(function()
+        WaitSeconds(2.9)
+        for _, v in GameOverListeners do
+            v()
+        end
+        Sync.GameEnded = true
+        WaitSeconds(0.1)
+        EndGame()
+    end)
+end
+
+--- Finds and collectors the brains that are defeated
+---@param aliveBrains AIBrain[] # Table of brains that are relevant to check for defeat
+---@param condition Categories  # Categories to check for units that are required to remain in the game
+---@param delay number          # Delay between each brain to spread the load over various ticks
+---@return AIBrain[]            # Table of brains that are considered defeated, can be empty
+local function CollectDefeatedBrains(aliveBrains, condition, delay)
+    local defeatedBrains = { }
+    for k, brain in aliveBrains do
+        local criticalUnits = brain:GetListOfUnits(condition)
+        if criticalUnits then
+            -- critical units found, make sure they all exist properly
+            local oneCriticalUnitAlive = false
+            for k, unit in criticalUnits do
```

Comment:
`k` not used in the loop? Also using `k` in the outer loop so this is a bit confusing.

```suggestion
            for _, unit in criticalUnits do
```